File: src/main/java/com/larryhsiao/clotho/stream/StreamString.java
package com.larryhsiao.clotho.stream;

import com.larryhsiao.clotho.Source;
import com.larryhsiao.clotho.io.ProgressedCopy;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;

/**
 * Source to build string from {@link InputStream}.
 */
public class StreamString implements Source<String> {
    private final InputStream inputStream;

    public StreamString(InputStream inputStream) {
        this.inputStream = inputStream;
    }

    @Override
    public String value() {
        try {
            final ByteArrayOutputStream byteArrayOutput = new ByteArrayOutputStream();
            new ProgressedCopy(
                inputStream,
                byteArrayOutput,
                1024,
                integer -> null
            ).fire();
            inputStream.close();
            byteArrayOutput.close();
            return byteArrayOutput.toString();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/java/com/silverhetch/clotho/stream/StreamString.kt
package com.silverhetch.clotho.stream

import com.silverhetch.clotho.Source
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.nio.charset.StandardCharsets.UTF_8

/**
 * Source to build [String] from given [InputStream]
 */
class StreamString(private val stream: InputStream) : Source<String> {
    override fun value(): String {
        val result = ByteArrayOutputStream()
        val buffer = ByteArray(4096)
        var length: Int
        while (stream.read(buffer).also { length = it } != -1) {
            result.write(buffer, 0, length)
        }
        return result.toString(UTF_8.name())
    }
}